src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -784,78 +794,132 @@ object Gen extends GenArities with GenVersionSpecific {
 
   //// Character Generators ////
 
+  private def charSample(cs: Array[Char]): Gen[Char] =
+    new Gen[Char] {
+      def doApply(p: P, seed0: Seed): Gen.R[Char] = {
+        val (x, seed1) = seed0.long
+        val i = ((x & Long.MaxValue) % cs.length).toInt
+        r(Some(cs(i)), seed1)
+      }
+    }
+
   /** Generates a numerical character */
-  def numChar: Gen[Char] = choose(48.toChar, 57.toChar)
+  val numChar: Gen[Char] =
+    charSample(('0' to '9').toArray)
 
   /** Generates an upper-case alpha character */
-  def alphaUpperChar: Gen[Char] = choose(65.toChar, 90.toChar)
+  val alphaUpperChar: Gen[Char] =
+    charSample(('A' to 'Z').toArray)
 
   /** Generates a lower-case alpha character */
-  def alphaLowerChar: Gen[Char] = choose(97.toChar, 122.toChar)
+  val alphaLowerChar: Gen[Char] =
+    charSample(('a' to 'z').toArray)
 
   /** Generates an alpha character */
-  def alphaChar = frequency((1,alphaUpperChar), (9,alphaLowerChar))
+  val alphaChar: Gen[Char] =
+    charSample((('A' to 'Z') ++ ('a' to 'z')).toArray)
 
   /** Generates an alphanumerical character */
-  def alphaNumChar = frequency((1,numChar), (9,alphaChar))
+  val alphaNumChar: Gen[Char] =
+    charSample((('0' to '9') ++ ('A' to 'Z') ++ ('a' to 'z')).toArray)
 
   /** Generates a ASCII character, with extra weighting for printable characters */
-  def asciiChar: Gen[Char] = chooseNum(0, 127, 32 to 126:_*).map(_.toChar)
+  val asciiChar: Gen[Char] =
+    choose(0, 127).map(_.toChar)
 
   /** Generates a ASCII printable character */
-  def asciiPrintableChar: Gen[Char] = choose(32.toChar, 126.toChar)
+  val asciiPrintableChar: Gen[Char] =
+    choose(32.toChar, 126.toChar)
 
   /** Generates a character that can represent a valid hexadecimal digit. This
     * includes both upper and lower case values.
     */
-  def hexChar: Gen[Char] =
-    Gen.oneOf(
-      Gen.oneOf("0123456789abcdef".toSeq),
-      Gen.oneOf("0123456789ABCDEF".toSeq)
-    )
+  val hexChar: Gen[Char] =
+    charSample("0123456789abcdef0123456789ABCDEF".toArray)
+
+  // valid ranges are [0x0000, 0xD7FF] and [0xE000, 0xFFFD].
+  //
+  // ((0xFFFD + 1) - 0xE000) + ((0xD7FF + 1) - 0x0000)
+  val unicodeChar: Gen[Char] =
+    choose(0, 63486).map { i =>
+      if (i <= 0xD7FF) i.toChar
+      else (i + 2048).toChar
+    }
 
   //// String Generators ////
 
+  @tailrec private def mkString(n: Int, sb: StringBuilder, gc: Gen[Char], p: P, seed0: Seed): R[String] =
+    if (n <= 0) {
+      r(Some(sb.toString), seed0)
+    } else {
+      val res = gc.doApply(p, seed0)
+      res.retrieve match {
+        case Some(c) =>
+          sb += c
+        case None =>
+          ()
```

Comment:
this is going to do a string of at most n, not exactly n, right? That's a change isn't it? Again, here is a place for a retry strategy.

Reply:
Yeah that's a good point.